Always send modifier release events, and clear focus on Move/Resize, window switcher and menu interactions

tokyo4j opened the pull request:
First, this PR reverts #1226 and #2377, because:
- Not sending modifier release events (#1226) for keybinds (for not showing firefox's menu bar) is very problematic with some applications (#1537, #1507).
- Forcing users to define `wantAbsorbedModifierReleaseEvents` window rule (#2377) to fix it is also undesirable, since it doesn't provide a good out-of-the-box experience and requires users to read through our documentation.

After some discussion in #2448 and #2452, we decided to follow KWin's approach that always sends modifier release events, but clears keyboard focus when activating window switcher (typically triggered by Alt-Tab) to prevent firefox from showing menu bar. Of course, this doesn't work for other actions like `Execute` and makes existing Alt-* keybinds show firefox's menu bar, but we concluded this approach is the most robust.

So in this PR, keyboard and pointer focus are temporarily cleared while Move/Resize, window switcher and menu interactions and restored after them, which basically follows KWin's behavior.

We slightly deviate from KWin as KWin doesn't clear the keyboard focus while Move/Resize, but it solves our existing problem that Firefox shows its menu bar after dragging it with default Alt-Drag mousebind, and this is what Mutter does.

Testing is appreciated.

tokyo4j commented:
"Temporarily clearing keyboard focus" in a reliable way is harder than I thought... I'll un-draft this PR when it's ready.

tokyo4j commented:
I think now this is ready. I'll daily-drive this PR this from today.

jlindgren90 commented:
The code looks sane to me. I have not tested it yet.

tokyo4j commented on include/labwc.h:
```diff
@@ -115,6 +115,14 @@ struct seat {
 		double x, y;
 	} smooth_scroll_offset;
 
+	/*
+	 * The surface whose keyboard focus is temporarily cleared with
+	 * seat_begin_server_action() and restored with
+	 * seat_finish_server_action().
+	 */
+	struct wlr_surface *last_focused_surface;
+	struct wl_listener last_focused_surface_destroy;
+
```
Note: Perhaps this can be substituted by `server->active_view` and `seat->focused_layer`, but an advantage of this approach is that it handles all types of surfaces including unmanaged-xwayland surfaces.

johanmalm replied:
IIRC `view->active_view` isn't always in sync either (e.g. when override redirect surfaces are focused). I might be getting my wires crossed on that one though.

jlindgren90 replied:
Just agreeing with both - I like the last_focused_surface approach as @tokyo4j has it. The active view vs. focused surface duality is annoying but I think we have to deal with it, because of the non-view surface types.

johanmalm commented on src/menu/menu.c:
```diff
@@ -1625,7 +1626,8 @@ menu_execute_item(struct menuitem *item)
 	menu_close(server->menu_current);
 	server->input_mode = LAB_INPUT_STATE_PASSTHROUGH;
-	cursor_update_focus(server);
+	seat_finish_server_action(&server->seat,
```
Should we consider setting `server->input_mode` in `seat_{begin,finish}_server_action()` functions? They might be set in other places - not looked - but it would feel like a neat thing to do.

tokyo4j replied:
Yeah you're absolutely right, and that's what I described as TODO in the header file. As I described there, I think the cleanest option is to add `LAB_INPUT_STATE_WINDOW_SWITCHER` in the `input_mode` enum and set it in those functions, but I didn't in this PR since it's kind of unpredictable how it affects other places like `input_mode != LAB_INPUT_STATE_PASSHROUGH` checks.

I'm thinking of opening another PR for just adding `LAB_INPUT_STATE_WINDOW_SWITCHER`.

tokyo4j replied:
Done, with help of #2459.

johanmalm commented on include/labwc.h:
```diff
@@ -115,6 +115,14 @@ struct seat {
+	 * seat_begin_server_action() and restored with
+	 * seat_finish_server_action().
+	 */
+	struct wlr_surface *last_focused_surface;
```
Consider wrapping these two variables in a `struct { } seat_server_mode` or something like that for ease of reading.

tokyo4j replied:
Good point. Wrapped with `focus_override`.

johanmalm reviewed (commented):
Have just left some quick comments for now.
Will use this PR for work tomorrow and then do a deeper review to follow the logic through.

jlindgren90 commented on src/seat.c:
```diff
@@ -794,3 +814,54 @@ seat_output_layout_changed(struct seat *seat)
+}
+
+void
+seat_begin_server_action(struct seat *seat, enum lab_cursors cursor_shape)
```
seat_focus_override_begin/end?

"action" and "mode" are both so heavily overloaded already...

tokyo4j replied:
Another option is `seat_set_input_mode()` and `seat_clear_input_mode()`, which depends on the change suggested in https://github.com/labwc/labwc/pull/2455#discussion_r1898728253 and #2459.

But I really like seat_focus_override_begin/end as it tells exactly what they do. I'll rename to it.

tokyo4j commented:
I added a condition in `cursor_update_common()` to ensure pointer focus is kept cleared while server-side interactions.

Also, I guess the following code is no longer needed, because it was a workaround to make clients update the cursor on finishing Move/Resize while keeping pointer focus on the client:

https://github.com/labwc/labwc/blob/1550eb9c26fcf756c3915e8ab9a777245c8f558c/src/input/cursor.c#L530-L548

@jlindgren90 @johanmalm Is my assumption right?

jlindgren90 commented:
Hmm, I am not sure about the pointer focus bits. I am less familiar with how that works.

tokyo4j commented:
I rebased this PR on #2459, and addressed the comments. I removed the part I quoted above.

johanmalm commented:
LGTM - other than rebasing against master now that #2459 has been merged.

Have run with it all day and it seems fine.

I suggest that we go for a merge soon. That way we're relatively early in the release cycle which gives us time to catch any unintended glitches.

Many thanks for working on this. It's really nice to get the fundamentals right.

johanmalm commented:
> Also, I guess `wlr_seat_pointer_notify_clear_focus()` in the following code is no longer needed, because it was a workaround to make clients update the cursor on finishing Move/Resize while keeping pointer focus on the client:

> Is my assumption right?

I think it's fine. Have had some git-blaming to see if those commits help job my memory, but can't think of anything else.

Getting all the keyboard-focus + cursor-updates right is quite finicky, but if we've missed some odd edge-case, we'll re-discover it and can document it better next time.

tokyo4j commented:
> I suggest that we go for a merge soon. That way we're relatively early in the release cycle which gives us time to catch any unintended glitches.

> Getting all the keyboard-focus + cursor-updates right is quite finicky, but if we've missed some odd edge-case, we'll re-discover it and can document it better next time.

Yeah I agree. I realized it's almost impossible to 100% make sure this doesn't cause any regression.

As I described as a `FIXME`, a known issue is that the focus is transiently given to the old surface at the end of window-switching. But I guess it won't cause any problems with most apps.

tokyo4j commented:
The changes in this PR will be described in the next release note like:

```
CHANGED:
- When a keybind is triggered while focusing on an application, the key-release event of the modifier key
  is now sent to the application. This fixes the problem of stuck modifier keys in some applications
  including Blender and wlfreerdp.
- Removed "wantAbsorbedModifierReleaseEvents" window rule as it's no longer needed
- Now the keyboard/pointer focus is cleared while a window is dragged, the window switcher is activated
  or a menu is opened.
```
